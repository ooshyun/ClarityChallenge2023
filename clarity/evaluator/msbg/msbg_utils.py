"""Support for the MSBG hearing loss model."""
import json
import logging
import math
import os
from typing import Dict, Optional, Tuple, Union

import numpy as np
import scipy
import scipy.signal
import soundfile
from soundfile import SoundFile

# measure rms parameters
WIN_SECS = 0.01

# read & write signal parameters
MSBG_FS = 44100
TEST_NBITS = 16

# fmt: off
HZ = np.array(
    [
        0.0, 20.0, 25.0, 31.5, 40.0, 50.0, 63.0, 80.0, 100.0, 125.0, 160.0, 200.0,
        250.0, 315.0, 400.0, 500.0, 630.0, 750.0, 800.0, 1000.0, 1250.0, 1500.0,
        1600.0, 2000.0, 2500.0, 3000.0, 3150.0, 4000.0, 5000.0, 6000.0, 6300.0,
        8000.0, 9000.0, 10000.0, 11200.0, 12500.0, 14000.0, 15000.0, 16000.0,
        20000.0, 48000,
    ]
)

MIDEAR = np.array(
    [
        50.0, 39.6, 32.0, 25.85, 21.4, 18.5, 15.9, 14.1, 12.4, 11.0, 9.6, 8.3, 7.4,
        6.2, 4.8, 3.8, 3.3, 2.9, 2.6, 2.6, 4.5, 5.4, 6.1, 8.5, 10.4, 7.3, 7.0, 6.6,
        7.0, 9.2, 10.2, 12.2, 10.8, 10.1, 12.7, 15.0, 18.2, 23.8, 32.3, 50.0, 50.0,
    ]
)

# Free field (frontal)FF_ED correction for threshold (was ISO std Table 1 - 4.2 dB)
# i.e. relative to 0.0 dB at 1000 Hz, Shaw 1974

FF_ED = np.array(
    [
        0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.1, 0.3, 0.5, 0.9, 1.4,
        1.6, 1.7, 2.5, 2.7, 2.6, 2.6, 3.2, 5.2, 6.6, 12.0, 16.8, 15.3, 15.2, 14.2,
        10.7, 7.1, 6.4, 1.8, -0.9, -1.6, 1.9, 4.9, 2.0, -2.0, 2.5, 2.5, 2.5,
    ]
)

# DIFFUSE field ( relative to 0.0 dB at 1000Hz)
# from 2008 file [corrections08.m] used in Samsung collaboration

DF_ED = np.array(
    [
        0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.1, 0.3, 0.4, 0.5, 1.0,
        1.6, 1.7, 2.2, 2.7, 2.9, 3.8, 5.3, 6.8, 7.2, 10.2, 14.9, 14.5, 14.4,
        12.7, 10.8, 8.9, 8.7, 8.5, 6.2, 5.0, 4.5, 4.0, 3.3, 2.6, 2.0, 2.0, 2.0,
    ]
)

# ITU Rec P 58 08/96 Head and Torso Simulator transfer fns. from Peter Hugher BTRL,
# 4-June-2001. Negative of values in Table 14a of ITU P58 (05/2013), accessible at
# http://www.itu.int/rec/T-REC-P.58-201305-I/en
# Freely available. Converts from ear reference point (ERP) to eardrum reference
# point (DRP). EXCEPT extra 2 points added for 20k & 48k by MAS, MAr 2012

ITU_HZ = np.array(
    [
        0, 100, 125, 160, 200, 250, 315, 400, 500, 630, 800, 1000, 1250, 1600,
        2000, 2500, 3150, 4000, 5000, 6300, 8000, 10000, 20000, 48000,
    ]
)
# Ear Reference Point to Drum Reference Point (ERP-DRP) transfer function,
# Table 14A/P.58, sect 6.2. NB negative of table since defined other way round.

ITU_ERP_DRP = np.array(
    [
        0.0, 0.0, 0.0, 0.0, 0.0, 0.3, 0.2, 0.5, 0.6, 0.7, 1.1, 1.7, 2.6, 4.2,
        6.5, 9.4, 10.3, 6.6, 3.2, 3.3, 16, 14.4, 14.4, 14.4
    ]
)
# fmt: on

# Ideal pre-emphasis (starts off as SII 1997), then rescaled for Moore et al. 2008
# NB last two are -15dB/oct before rescaling below
# (Moore et al 2008 E&H paper suggests that shape would be better as -7.5
# dB/oct, at least up to 8, and -13 dB/oct above there.)

GEN_NOISE_HZ = np.array(
    [0, 100, 200, 450, 550, 707, 1000, 1414, 2000, 2828, 4000, 5656, 8000, 16e3, 32e3]
)
EMPHASIS = np.array(
    [0, 0.0, 0.0, 0, -0.5, -4.5, -9.0, -13.5, -18, -22.5, -27, -31.5, -36.0, -51, -66]
) * (7.5 / 9)


def read_gtf_file(gtf_file: str) -> Dict:
    """Read a gammatone filterbank file.

    List data is converted into numpy arrays.

    """

    # Fix filename if necessary
    dirname = os.path.dirname(os.path.abspath(__file__))
    gtf_file = os.path.join(dirname, gtf_file)
    with open(gtf_file, "r", encoding="utf-8") as fp:
        data = json.load(fp)
    for key in data:
        if isinstance(data[key], list):
            data[key] = np.array(data[key])
    return data


def firwin2(
    n_taps: int,
    frequencies: np.ndarray,
    filter_gains: np.ndarray,
    window: Union[None, str, tuple] = None,
    antisymmetric: Optional[bool] = None,
):  # pylint: disable=W0613
    """FIR filter design using the window method.

    Partial implementation of scipy firwin2 but using our own MATLAB-derived fir2.

    Args:
        n_taps (int): The number of taps in the FIR filter.
        frequencies (ndarray): The frequency sampling points. 0.0 to 1.0 with 1.0 being Nyquist.
        filter_gains (ndarray): The filter gains at the frequency sampling points.
        window (string or (string, float), optional): See scipy.firwin2 (default: (None))
        antisymmetric (bool, optional): Unused but present to main compatability
            with scipy firwin2.

    Returns:
        ndarray:  The filter coefficients of the FIR filter, as a 1-D array of length n.

    """
    window_shape = None
    window_type = None
    if isinstance(window, tuple):
        window_type, window_param = window if window is not None else (None, 0)
    else:
        window_type = window

    order = n_taps - 1

    if window_type == "kaiser":
        window_shape = scipy.signal.kaiser(n_taps, window_param)

    if window_shape is None:
        filter_coef, _ = fir2(order, frequencies, filter_gains)
    else:
        filter_coef, _ = fir2(order, frequencies, filter_gains, window_shape)

    return filter_coef


def fir2(
    filter_length: int,
    frequencies: np.ndarray,
    filter_gains: np.ndarray,
    n_interpolate: Optional[int] = None,
) -> Tuple[np.ndarray, int]:
    """FIR arbitrary shape filter design using the frequency sampling method.

    Translation of MATLAB fir2.

    Args:
        filter_length (int): Order
        frequencies (ndarray): The frequency sampling points (0 < frequencies < 1) where 1 is Nyquist rate.
                        First and last elements must be 0 and 1 respectively.
        filter_gains (ndarray): The filter gains at the frequency sampling points.
        n_interpolate (int, optional): Number of points for freq response interpolation
            (default: max(smallest power of 2 greater than nn, 512))

    Returns:
        np.ndarray: nn + 1 filter coefficients, 1

    """
    # Work with filter length instead of filter order
    filter_length += 1

    if n_interpolate is None:
        wind = scipy.signal.hamming(filter_length)
    else:
        wind = n_interpolate
    n_interpolate = (
        2 ** np.ceil(math.log(filter_length) / math.log(2))
        if filter_length >= 1024
        else 512
    )

    lap = np.fix(n_interpolate / 25).astype(int)

    nbrk = max(len(frequencies), len(filter_gains))

    frequencies[0] = 0
    frequencies[nbrk - 1] = 1

    H = np.zeros(n_interpolate + 1)
    nint = nbrk - 1
    df = np.diff(frequencies, n=1)

    n_interpolate += 1
    nb = 0
    H[0] = filter_gains[0]

    for i in np.arange(nint):
        if df[i] == 0:
            nb = int(np.ceil(nb - lap / 2))
            ne: int = nb + lap - 1
        else:
            ne = int(np.fix(frequencies[i + 1] * n_interpolate)) - 1

        j = np.arange(nb, ne + 1)
        inc: Union[float, np.ndarray] = 0.0 if nb == ne else (j - nb) / (ne - nb)
        H[nb : (ne + 1)] = inc * filter_gains[i + 1] + (1 - inc) * filter_gains[i]
        nb = ne + 1

    dt = 0.5 * (filter_length - 1)
    rad = -dt * 1j * math.pi * np.arange(0, n_interpolate) / (n_interpolate - 1)
    H = H * np.exp(rad)

    H = np.concatenate((H, H[n_interpolate - 2 : 0 : -1].conj()))
    ht = np.real(np.fft.ifft(H))

    b = ht[0:filter_length] * wind

    return b, 1


def gen_tone(
    freq: float, duration: float, sample_frequency: float = 44100.0, level: float = 0.0
) -> np.ndarray:
    """Generate a pure tone.

    Args:
        freq (float): Frequency of tone in Hz.
        duration (float): Duration of tone in seconds.
        sample_frequency (float, optional): Sample rate of generated tone in Hz. (default: 44100)
        level (float, optional): Level of tone in dB SPL. (default: 0)

    Returns:
        np.ndarray
    """
    return (
        1.4142
        * np.power(10, (0.05 * level))
        * np.sin(
            2
            * np.pi
            * freq
            * np.arange(1, duration * sample_frequency + 1)
            / sample_frequency
        )
    )


def gen_eh2008_speech_noise(
    duration: float,
    sample_frequency: float = 44100.0,
    level: Optional[float] = None,
    supplied_b: Optional[np.ndarray] = None,
) -> np.ndarray:
    """Generate speech shaped noise.

    Start with white noise and re-shape to ideal SII, ie flat to 500 Hz, and sloping -9db/oct beyond that.

    Slightly different shape from SII stylised same as EarHEar 2008 paper, Moore et al.

    Args:
        duration (float): Duration of signal in seconds
        sample_frequency (float): Sampling rate
        level (float, optional): Normalise to level dB if present
        supplied_b (ndarray, optional): High-pass filter (default: uses built-in pre-emphasis filter)

    Returns:
        ndarray: Noise signal

    """
    sample_frequency = int(sample_frequency)
    n_samples = int(duration * sample_frequency)

    # this rescales so that we get -7.5 dB/oct up to 8kHz, and -13 dB/oct above that
    norm_freq = GEN_NOISE_HZ / (sample_frequency / 2)
    last_f_idx = np.max(np.where(norm_freq < 1))
    norm_freq = np.append(norm_freq[0 : last_f_idx + 1], 1)

    # -9 dB/oct constant slope
    emph_nyq = EMPHASIS[last_f_idx] + 9 * np.log10(norm_freq[last_f_idx]) / np.log10(2)
    norm_emph = np.append(EMPHASIS[0 : last_f_idx + 1], emph_nyq)
    m = np.exp(np.log(10) * norm_emph / 20)

    # Create type II filter with 10 msec window and even number of taps
    n_taps = int(2 * np.ceil(10 * (sample_frequency / 2000))) + 1
    b = (
        supplied_b
        if supplied_b is not None
        else firwin2(n_taps, norm_freq, m, window="hamming", antisymmetric=False)
    )

    # white noise, 0 DC
    n_burst = np.random.random((1, n_samples + len(b))) - 0.5

    # remove low-freq noise that may bias RMS estimate, -33dB at 50 Hz
    eh2008_nse = scipy.signal.lfilter(b, 1, n_burst)

    # high-pass filter to remove low freqs (will be 2-pass with filtfilt)
    high_pass_filter = scipy.signal.ellip(
        3, 0.1, 50, 100 / (sample_frequency / 2), "high"
    )
    padlen = 3 * (max(len(high_pass_filter[1]), len(high_pass_filter[0])) - 1)
    eh2008_nse = scipy.signal.filtfilt(
        *high_pass_filter, eh2008_nse, padlen=padlen
    ).flatten()

    # this introduces a delay so remove it, ie time-ADVANCE audio
    # compensating shift to time-align all filter outputs
    delay_shift = int(np.floor(len(b) / 2))
    valid_len = int(np.size(eh2008_nse) - delay_shift)  # _advance_ filter outputs
    # time advance
    eh2008_nse[0:valid_len] = eh2008_nse[delay_shift:]
    eh2008_nse = eh2008_nse[0:n_samples]

    if level is not None:
        eh2008_nse = (
            eh2008_nse
            * np.power(10, 0.05 * level)
            / np.sqrt(np.sum(np.power(eh2008_nse, 2)) / len(eh2008_nse))
        )

    return eh2008_nse


def generate_key_percent(
    signal: np.ndarray,
    threshold_db: float,
    window_length: int,
    percent_to_track: Optional[float] = None,
) -> Tuple[np.ndarray, float]:
    """Generate key percent.
    Locates frames above some energy threshold or tracks a certain percentage
    of frames. To track a certain percentage of frames in order to get measure
    of rms, adaptively sets threshold after looking at histogram of whole recording

    Args:
        signal (ndarray): The signal to analyse.
        threshold_db (float): fixed energy threshold (dB).
        window_length (int): length of window in samples.
        percent_to_track (float, optional): Track a percentage of frames (default: {None}).

    Raises:
        ValueError: percent_to_track is set too high.

    Returns:
        (tuple): containig
        - key (ndarray): The key array of indices of samples used in rms calculation.
        - used_threshold_db (float): Root Mean Squared threshold.

            and the threshold used to get a more accurate rms calculation
    """
    window_length = int(window_length)
    signal = signal.flatten()
    if window_length != math.floor(
        window_length
    ):  # whoops on fractional indexing: 7-March 2002
        window_length = math.floor(window_length)
        logging.warning(f"Window length must be integer: now {window_length}")

    signal_length = len(signal)

    expected = threshold_db
    # new Dec 2003. Possibly track percentage of frames rather than fixed threshold
    if percent_to_track is not None:
        logging.info("tracking %s percentage of frames", percent_to_track)
    else:
        logging.info("tracking fixed threshold")

    # put floor into histogram distribution
    non_zero = np.power(10, (expected - 30) / 10)

    n_frames = -1
    total_frames = math.floor(signal_length / window_length)
    every_db = np.zeros(total_frames)

    for ix in np.arange(
        0, window_length * total_frames - 1, window_length
    ):  # pylint: disable=invalid-name
        n_frames += 1
        this_sum = np.sum(
            np.power(signal[ix : (ix + window_length)].astype("float"), 2)
        )
        every_db[n_frames] = 10 * np.log10(non_zero + this_sum / window_length)
    n_frames += 1

    # from now on save only those analysed
    every_db = every_db[:n_frames]

    # Bec 2003, was 100 to give about a 0.5 dB quantising of levels
    n_bins, levels = np.histogram(every_db, 140)
    if percent_to_track is not None:
        # min number of bins to use
        inactive_bins = (100 - percent_to_track) * n_frames / 100
        n_levels = len(levels)
        inactive_ix = 0
        ix_count = 0
        for ix in np.arange(0, n_levels, 1):  # pylint: disable=invalid-name
            inactive_ix = inactive_ix + n_bins[ix]
            if inactive_ix > inactive_bins:
                break
            ix_count += 1
        if ix == 1:
            logging.warning("Counted every bin.........")
        elif ix == n_levels:
            raise ValueError("Generate_key_percent: no levels to count")
        expected = levels[max(1, ix_count)]

    # set new threshold conservatively to include more bins than desired
    used_threshold_db = expected

    # histogram should produce a two-peaked curve: thresh should be set in valley
    # between the two peaks, and set threshold a bit above that,
    # as it heads for main peak
    # FixMe : Could Otsu's method (from image processing) be used here? https://en.wikipedia.org/wiki/Otsu's_method
    frame_index = np.nonzero(every_db >= expected)[0]
    valid_frames = len(frame_index)
    key = np.zeros((1, valid_frames * window_length))[0]

    # convert frame numbers into indices for sig
    for ix in np.arange(valid_frames):  # pylint: disable=invalid-name
        meas_span = np.arange(
            (frame_index[ix] * window_length), (frame_index[ix] + 1) * window_length
        )
        key_span = np.arange(((ix) * window_length), (ix + 1) * window_length, 1)
        key[key_span] = meas_span
        key = key.flatten()

    return key, used_threshold_db


def measure_rms(
    signal: np.ndarray,
    sample_frequency: float,
    db_rel_rms: float,
    percent_to_track: Optional[float] = None,
) -> tuple:
    """Measure Root Mean Square.

    A sophisticated method of measuring RMS in a file. It splits the signal up into
    short windows, performs  a histogram of levels, calculates an approximate RMS,
    and then uses that RMS to calculate a threshold level in the histogram and then
    re-measures the RMS only using those durations whose individual RMS exceed that
    threshold.

    Args:
        signal (ndarray): the signal of which to measure the Root Mean Square.
        sample_frequency (float): sampling frequency.
        db_rel_rms (float): threshold for frames to track.
        percent_to_track (float, optional): track percentage of frames,
            rather than threshold (default: {None})
    Returns:
        (tuple): tuple containing
        - rms (float): overall calculated rms (linear)
        - key (ndarray): "key" array of indices of samples used in rms calculation
        - rel_db_thresh (float): fixed threshold value of -12 dB
        - active (float): proportion of values used in rms calculation
    """
    sample_frequency = int(sample_frequency)
    # first RMS is of all signal.
    first_stage_rms = np.sqrt(np.sum(np.power(signal, 2) / len(signal)))
    # use this RMS to generate key threshold to get more accurate RMS
    key_thr_db = max(20 * np.log10(first_stage_rms) + db_rel_rms, -80)

    # move key_thr_db to account for noise less peakier than signal
    key, used_thr_db = generate_key_percent(
        signal,
        key_thr_db,
        round(WIN_SECS * sample_frequency),
        percent_to_track=percent_to_track,
    )

    idx = key.astype(int)  # move into generate_key_percent
    # statistic to be reported later, BUT save for later
    # (for independent==1 loop where it sets a target for rms measure)
    active = 100 * len(key) / len(signal)
    rms = np.sqrt(np.sum(np.power(signal[idx], 2)) / len(key))
    rel_db_thresh = used_thr_db - 20 * np.log10(rms)

    return rms, idx, rel_db_thresh, active


def pad(signal, length):
    """Zero pad signal to required length.

    Assumes required length is not less than input length.
    """
    assert length >= signal.shape[0]
    return np.pad(
        signal, [(0, length - signal.shape[0])] + [(0, 0)] * (len(signal.shape) - 1)
    )


def read_signal(
    filename: str,
    offset: int = 0,
    nsamples: int = -1,
    nchannels: int = 0,
    offset_is_samples: bool = False,
) -> np.ndarray:
    """Read a wavefile and return as numpy array of floats.

    Args:
        filename (string): Name of file to read
        offset (int, optional): Offset in samples or seconds (from start). Defaults to 0.
        nsamples (int): Number of samples.
        nchannels (int): expected number of channel (default: 0 = any number OK)
        offset_is_samples (bool): measurement units for offset (default: False)

    Returns:
        np.ndarray: audio signal
    """
    try:
        wave_file = SoundFile(filename)
    except Exception as e:
        # Ensure incorrect error (24 bit) is not generated
        raise Exception(f"Unable to read {filename}.") from e

    if nchannels not in (0, wave_file.channels):
        raise Exception(
            f"Wav file ({filename}) was expected to have {nchannels} channels."
        )

    if not offset_is_samples:  # Default behaviour
        offset = int(offset * wave_file.samplerate)

    if offset != 0:
        wave_file.seek(offset)

    signal = wave_file.read(frames=nsamples)

    if wave_file.samplerate != MSBG_FS:
        signal = scipy.signal.resample(
            signal, int(MSBG_FS * signal.shape[0] / wave_file.samplerate)
        )

    return signal


def write_signal(
    filename: str,
    signal: np.ndarray,
    sample_frequency: float,
    floating_point: bool = True,
) -> None:
    """Write a signal as fixed or floating point wav file.

    Args:
        filename (str): name of file in to write to.
        signal (ndarray): signal to write.
        sample_frequency (float): sampling frequency.
        floating_point (bool): write as floating point else an ints (default: True).
    """

    if sample_frequency != MSBG_FS:
        logging.warning(
            f"Sampling rate mismatch: {filename} with sample frequency = {sample_frequency}."
        )
        # raise ValueError("Sampling rate mismatch")

    if floating_point is False:
        if TEST_NBITS == 16:
            subtype = "PCM_16"
            # If signal is float and we want int16
            signal *= 32768
            signal = signal.astype(np.dtype("int16"))
            assert np.max(signal) <= 32767 and np.min(signal) >= -32768
        elif TEST_NBITS == 24:
            subtype = "PCM_24"
    else:
        subtype = "FLOAT"

    soundfile.write(filename, signal, sample_frequency, subtype=subtype)
