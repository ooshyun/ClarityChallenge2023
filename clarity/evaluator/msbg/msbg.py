"""Implementation of the MSBG hearing loss model."""
import logging
import math

import numpy as np
from scipy import interpolate, signal
from scipy.signal import firwin, lfilter

from clarity.evaluator.msbg.cochlea import Cochlea
from clarity.evaluator.msbg.msbg_utils import (
    DF_ED,
    FF_ED,
    HZ,
    ITU_ERP_DRP,
    ITU_HZ,
    MIDEAR,
    firwin2,
    gen_eh2008_speech_noise,
    gen_tone,
    measure_rms,
)

# Cut off frequency of low-pass filter at end of simulations:
# prevents possible excessive processing noise at high frequencies.
UPPER_CUTOFF_HZ = 18000


class Ear:
    """Representation of a pairs of ears."""

    def __init__(
        self, src_pos="ff", sample_frequency: float = 44100.0, equiv_0db_spl=100, ahr=20
    ) -> None:
        """
        Constructor for the Ear class.
        Args:
            src_pos (str): Position of the source.
            sample_frequency (float): sample frequency.
            equiv_0db_spl (): ???
            ahr (): ???
        """
        self.sample_frequency = sample_frequency
        self.src_correction = self.get_src_correction(src_pos)
        self.equiv_0db_spl = equiv_0db_spl
        self.ahr = ahr
        self.cochlea = None

    def set_audiogram(self, audiogram):
        """Set the audiogram to be used."""
        if np.max(audiogram.levels[audiogram.levels is not None]) > 80:
            logging.warning(
                "Impairment too severe: Suggest you limit audiogram max to 80-90 dB HL, \
                otherwise things go wrong/unrealistic."
            )
        self.cochlea = Cochlea(audiogram=audiogram)

    @staticmethod
    def get_src_correction(src_pos: str) -> np.array:
        """Select relevant external field to eardrum correction.

        Args:
            src_pos (str): Position of src. One of ff, df or ITU

        """
        if src_pos == "ff":
            src_correction = FF_ED
        elif src_pos == "df":
            src_correction = DF_ED
        elif src_pos == "ITU":  # transfer to same grid
            field = interpolate.interp1d(ITU_HZ, ITU_ERP_DRP, kind="linear")
            src_correction = field(HZ)
        else:
            logging.error(
                f"Invalid src position ({src_pos}). Must be one of ff, df or ITU"
            )
            raise ValueError("Invalid src position")
        return src_correction

    @staticmethod
    def src_to_cochlea_filt(
        ip_sig: np.ndarray, src_correction: str, sample_frequency: int, backward=False
    ) -> np.ndarray:
        """Simulate middle and outer ear transfer functions.

        Made more general, Mar2012, to include diffuse field as well as ITU reference points,
        that were included in DOS-versions of recruitment simulator, released ca 1999-2001,
        and on hearing group website, Mar2012 variable [src_pos] takes one of 3
        values: 'ff', 'df' and 'ITU' free-field to cochlea filter forwards or backward direction,
        depends on 'backward' switch. NO LONGER via 2 steps. ff to eardrum and then via middle ear:
        use same length FIR 5-12-97.

        Args:
            ip_sig (ndarray): signal to process
            src_correction (str): correction to make for src position
            sample_frequency (int): sampling frequency
            backward (bool, optional): if true then cochlea to src (default: False)

        Returns:
            np.ndarray: the processed signal

        """
        logging.info("performing outer/middle ear corrections")

        # make sure that response goes only up to sample_frequency/2
        nyquist = int(sample_frequency / 2)
        ixf_useful = np.nonzero(HZ < nyquist)

        hz_used = HZ[ixf_useful]
        hz_used = np.append(hz_used, nyquist)

        # sig from free field to cochlea: 0 dB at 1kHz
        correction = src_correction - MIDEAR
        field = interpolate.interp1d(HZ, correction, kind="linear")
        last_correction = field(nyquist)  # generate synthetic response at Nyquist

        correction_used = np.append(correction[ixf_useful], last_correction)
        if backward:  # ie. coch->src rather than src->coch
            correction_used = -correction_used
        correction_used = np.power(10, (0.05 * correction_used))

        correction_used = correction_used.flatten()
        # Create filter with 23 msec window to do reasonable job down to about 100 Hz
        # Scales with fs, falls over with longer windows in fir2 in original MATLAB version
        n_wdw = 2 * math.floor((sample_frequency / 16e3) * 368 / 2)
        hz_used = hz_used / nyquist

        b = firwin2(n_wdw + 1, hz_used.flatten(), correction_used, window=("kaiser", 4))
        op_sig = signal.lfilter(b, 1, ip_sig)

        return op_sig

    def make_calibration_signal(self, ref_rms_db):
        """Add the calibration signal to the start of the signal.

        Args:
            ref_rms_db (ndarray): input signal

        Returns:
            ndarray: the processed signal

        """
        # Calibration noise and tone with same RMS as original speech,
        # Tone at nearest channel centre frequency to 500 Hz
        # For testing, ref_rms_dB must be equal to -31.2

        noise_burst = gen_eh2008_speech_noise(
            duration=2, sample_frequency=self.sample_frequency, level=ref_rms_db
        )
        tone_burst = gen_tone(
            freq=520,
            duration=0.5,
            sample_frequency=self.sample_frequency,
            level=ref_rms_db,
        )
        silence = np.zeros(int(0.05 * self.sample_frequency))  # 50 ms duration
        return (
            np.concatenate((silence, tone_burst, silence, noise_burst, silence)),
            silence,
        )

    @staticmethod
    def array_to_list(chans: np.ndarray) -> np.ndarray:
        """Convert signal into a list of 1-D arrays.

        Args:
            signal (np.ndarray) Signal to be converted, can be 1-D already.

        Returns:
            np.ndarray: A list of 1-D arrays.
        """
        if len(chans.shape) == 1:
            chans = chans[..., np.newaxis]
        return [chans[:, i] for i in range(chans.shape[1])]

    def process(self, signal: np.ndarray, add_calibration: bool = False) -> np.ndarray:
        """Run the hearing loss simulation.

        Args:
            signal (ndarray): signal to process, shape either N, Nx1, Nx2
            add_calibration (bool): prepend calibration tone and speech-shaped noise
                (default: False)

        Returns:
            np.ndarray: the processed signal

        """

        sample_frequency = 44100  # This is the only sampling frequency that can be used
        if sample_frequency != self.sample_frequency:
            logging.error(
                "Warning: only a sampling frequency of 44.1kHz can be used by MSBG."
            )

        logging.info("Processing {len(chans)} samples")

        # Get single channel array and convert to list
        signal = Ear.array_to_list(signal)

        # Need to know file RMS, and then call that a certain level in SPL:
        # needs some form of pre-measuring.
        level_resample_frequency = 10 * np.log10(np.mean(np.array(signal) ** 2))

        equiv_0db_spl = self.equiv_0db_spl + self.ahr

        level_db_spl = equiv_0db_spl + level_resample_frequency
        calib_db_spl = level_db_spl
        target_spl = level_db_spl
        ref_rms_db = calib_db_spl - equiv_0db_spl

        # Measure RMS where 3rd arg is dB_rel_rms (how far below)
        calculated_rms, idx, _rel_db_thresh, _active = measure_rms(
            signal[0], sample_frequency, -12
        )

        # Rescale input data and check level after rescaling
        # This is to ensure that the following processing steps are applied correctly
        change_db = target_spl - (equiv_0db_spl + 20 * np.log10(calculated_rms))
        signal = [x * np.power(10, 0.05 * change_db) for x in signal]
        new_rms_db = equiv_0db_spl + 10 * np.log10(
            np.mean(np.power(signal[0][idx], 2.0))
        )
        logging.info(
            "Rescaling: "
            f"leveldBSPL was {level_db_spl:3.1f} dB SPL, now {new_rms_db:3.1f} dB SPL. "
            f" Target SPL is {target_spl:3.1f} dB SPL."
        )

        # Add calibration signal at target SPL dB
        if add_calibration is True:
            # if self.calibration_signal is None:
            #     self.calibration_signal = self.make_calibration_signal(ref_rms_db)
            # signal = [
            #     np.concatenate(
            #         (self.calibration_signal[0], x, self.calibration_signal[1])
            #     )
            #     for x in signal
            # ]
            calibration_signal = self.make_calibration_signal(ref_rms_db)
            signal = [
                np.concatenate((calibration_signal[0], x, calibration_signal[1]))
                for x in signal
            ]

        # Transform from src pos to cochlea, simulate cochlea, transform back to src pos
        signal = [
            Ear.src_to_cochlea_filt(x, self.src_correction, sample_frequency)
            for x in signal
        ]
        if self.cochlea is not None:
            signal = [self.cochlea.simulate(x, equiv_0db_spl) for x in signal]
        signal = [
            Ear.src_to_cochlea_filt(
                x, self.src_correction, sample_frequency, backward=True
            )
            for x in signal
        ]

        # Implement low-pass filter at top end of audio range: flat to Cutoff freq, tails
        # below -80 dB. Suitable lpf for signals later converted to MP3, flat to 15 kHz.
        # Small window to design low-pass FIR, to cut off high freq processing noise
        # low-pass to something sensible, prevents exaggeration of > 15 kHz
        winlen = 2 * math.floor(0.0015 * sample_frequency) + 1
        lpf44d1 = firwin(
            winlen, UPPER_CUTOFF_HZ / int(sample_frequency / 2), window=("kaiser", 8)
        )
        signal = [lfilter(lpf44d1, 1, x) for x in signal]

        return signal
