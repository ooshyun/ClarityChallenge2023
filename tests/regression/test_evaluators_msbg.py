import numpy as np
import pytest

from clarity.evaluator.msbg.msbg_utils import (
    firwin2,
    gen_eh2008_speech_noise,
    gen_tone,
    measure_rms,
    pad,
)


@pytest.fixture
def use_numpy():
    np.random.seed(0)
    np_print_opts = np.get_printoptions()
    np.set_printoptions(precision=5, threshold=1000)
    yield
    np.set_printoptions(**np_print_opts)


def test_gen_eh2008_speech_noise(regtest, use_numpy):
    signal = gen_eh2008_speech_noise(0.1, 44100.0, 0.0)
    with regtest:
        print(signal[-9:])


def test_gen_tone(regtest, use_numpy):
    signal = gen_tone(500, 0.1, 44100.0, 20.0)
    with regtest:
        print(signal[-9:])


def test_firwin2(regtest, use_numpy):
    params = firwin2(128, [0.0, 0.1, 0.9, 1.0], [0, 1.0, 1.0, 0.0])
    with regtest:
        print(params)


def test_pad(regtest, use_numpy):
    padded = pad(np.array([1.0, 1.0, 1.0, 1.0]), 100)
    with regtest:
        print(padded)


def test_measure_rms(use_numpy, regtest):
    signal = gen_tone(500, 0.05, 44100.0, 20.0)
    noise = gen_eh2008_speech_noise(0.05, 44100.0, 0.0)
    rms, idx, rel_dB_thresh, active = measure_rms(signal + noise, 44100, 0.0, 10.0)
    with regtest:
        print(f"{rms:.7f}, {idx}, {rel_dB_thresh:.7f}, {active:.7f}")
