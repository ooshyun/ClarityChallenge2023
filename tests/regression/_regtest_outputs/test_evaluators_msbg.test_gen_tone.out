[-7.63033e+00 -6.76347e+00 -5.86230e+00 -4.93139e+00 -3.97547e+00
 -2.99939e+00 -2.00808e+00 -1.00659e+00 -7.76099e-13]
